### PR TITLE
[NVPTX] Load/Store/Fence syncscope support

### PR DESCRIPTION
**Pull request by gonzalobg**
Adds "initial" support for `syncscope` to the NVPTX backend `load`/`store`/`fence` instructions.
Atomic Read-Modify-Write operations intentionally not supported as part of this initial PR.

### CODE REVIEW

**Review by AlexMaclean — APPROVED** [review]
LGTM, thanks

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 1]
```diff
@@ -4064,3 +4158,43 @@ unsigned NVPTXDAGToDAGISel::GetConvertOpcode(MVT DestTy, MVT SrcTy,
...
+
+  auto S = Scopes.find(ID);
+  if (S == Scopes.end()) {
+    SmallVector<StringRef, 8> ScopeNames;
+    assert(CTX != nullptr && "CTX is nullptr");
+    CTX->getSyncScopeNames(ScopeNames);
+    StringRef Unknown{"unknown"};
+    auto Name = ID < ScopeNames.size() ? ScopeNames[ID] : Unknown;
+    report_fatal_error(
+        formatv("Could not find scope ID={} with name \"{}\".", int(ID), Name));
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4195]: I don't think we're that interested in the name of the scope here. Just report the failing ID value and move on.

If you do want the name, a better way to do it would be to add an API to return the name for the given ID (that would go into a separate patch). There's no point copying all the names just to find one.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 2]
```diff
@@ -915,6 +920,42 @@ getOperationOrderings(MemSDNode *N, const NVPTXSubtarget *Subtarget) {
 
 } // namespace
 
+NVPTX::Scope NVPTXDAGToDAGISel::getOperationScope(MemSDNode *N,
+                                                  NVPTX::Ordering Ord) const {
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:924]: Nit: `O` or `Ordering` would work a bit better here, IMO. I'd probably use just `O` to match `N`.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 3]
```diff
@@ -4064,3 +4158,43 @@ unsigned NVPTXDAGToDAGISel::GetConvertOpcode(MVT DestTy, MVT SrcTy,
...
+  Scopes[C.getOrInsertSyncScopeID("singlethread")] = NVPTX::Scope::Thread;
+  Scopes[C.getOrInsertSyncScopeID("")] = NVPTX::Scope::System;
+  Scopes[C.getOrInsertSyncScopeID("block")] = NVPTX::Scope::Block;
+  Scopes[C.getOrInsertSyncScopeID("cluster")] = NVPTX::Scope::Cluster;
+  Scopes[C.getOrInsertSyncScopeID("device")] = NVPTX::Scope::Device;
+}
+
+NVPTX::Scope NVPTXScopes::operator[](SyncScope::ID ID) const {
+  if (Scopes.empty())
+    report_fatal_error("NVPTX Scopes must be initialized before calling "
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4184]: This should be an assert or `llvm_unreachable`, as the failure to init the scopes info is not a user error.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h [hunk 4]
```diff
@@ -100,8 +116,9 @@ class LLVM_LIBRARY_VISIBILITY NVPTXDAGToDAGISel : public SelectionDAGISel {
 
   static unsigned GetConvertOpcode(MVT DestTy, MVT SrcTy, LoadSDNode *N);
 
-  NVPTX::Ordering insertMemoryInstructionFence(SDLoc DL, SDValue &Chain,
-                                               MemSDNode *N);
+  std::pair<NVPTX::Ordering, NVPTX::Scope>
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:119]: A comment describing what returned `Ordering` and `Scope` have to do with the instruction insertion would be helpful.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h [hunk 5]
```diff
@@ -18,13 +18,26 @@
...
 namespace llvm {
 
+struct NVPTXScopes {
+  NVPTXScopes() = default;
+  NVPTXScopes(LLVMContext &C);
+  NVPTX::Scope operator[](SyncScope::ID ID) const;
+  bool empty() const;
+
+private:
+  SmallMapVector<SyncScope::ID, NVPTX::Scope, 8> Scopes{};
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:37]: We have exactly 5 scopes here, so we can trim the small vector size down, too. Maybe define enum value for the number of scopes and use that.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 6]
```diff
@@ -1189,7 +1283,7 @@ bool NVPTXDAGToDAGISel::tryLoad(SDNode *N) {
   std::optional<unsigned> Opcode;
   MVT::SimpleValueType TargetVT = LD->getSimpleValueType(0).SimpleTy;
 
-  SmallVector<SDValue, 12> Ops({getI32Imm(InstructionOrdering, DL),
+  SmallVector<SDValue, 12> Ops({getI32Imm(Ordering, DL), getI32Imm(Scope, DL),
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1286]: We've added another element to the array. Does it bump the max number of elements ever used past 12? If so, vector size may need to be increased.

File: llvm/lib/Target/NVPTX/NVPTXSubtarget.h [hunk 7]
```diff
@@ -118,6 +119,8 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
 
   NVPTXSubtarget &initializeSubtargetDependencies(StringRef CPU, StringRef FS);
   void ParseSubtargetFeatures(StringRef CPU, StringRef TuneCPU, StringRef FS);
+
+  void requireClusters(std::string const &FailureMessage) const;
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXSubtarget.h:123]: We need to make clear that this produces a fatal error if conditions are not met.

`require` alone does not carry "I am a fatal error" vibe.  
I think we need something painfully obvious, along the lines of `failIfClustersNotAvailable()`.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h (hunk 5 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:37]: SmallMapVector only accepts power of two Capacities (so either 4 or 8, but not 5).

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 6 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1286]: 12 is just the capacity, so the worst that may happen is that it falls back to heap allocation.
If we want a vector that fails when the static capacity is exceeded (to prevent any heap allocation fallback), using C++26 `inplace_vector<T, Capacity>` is probably a better choice. If that's the case, does LLVM have something like that?

File: llvm/lib/Target/NVPTX/NVPTX.h [hunk 8]
```diff
@@ -117,12 +117,22 @@ enum Ordering : OrderingUnderlyingType {
...
+};
+
+using ScopeUnderlyingType = unsigned int;
+enum Scope : ScopeUnderlyingType {
+  Thread = 0,
+  System = 1,
+  Block = 2,
+  Cluster = 3,
+  Device = 4,
+  LASTSCOPE = Device
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTX.h:135]: I think that in retrospect, it does.

Currently, this extends the LLVM SyncScope::IDs, which reserve 0 and 1 for system and thread scopes, in a similar way to how NVPTX Ordering extends LLVM AtomicOrdering. 

The value of doing that is that you can just cast the LLVM value to our value, when processing the DAG, and that the `i32 3` in the DAG have the same meaning in LLVM enum as in NVPTX enum. For atomic ordering, this is quite useful.

For SyncScope, this is only useful for system scope. For the other SyncScopes, their value in the DAG is dynamic and depends on which order `syncscope`s where used in the LLVM module.. so to make sense of them one needs to do a translation to NVPTX::Scope via the dynamic map anyways. That is, extending the LLVM enum order is much less useful for SyncScope, than for AtomicOrdering.

I think that it is worth trying out whether ordering these from narrower to wider scope (thread =0 , system=last) allows us to save some code here and there (e.g. `if scope > thread: ...`). That's a non functional change, and it may or may not help (we may still need full `switch` statements here and there, cause some of the scopes are sm-arch dependent...). I think its worth giving that a try. Should I try to do it here, or in a follow up?

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp [hunk 9]
```diff
@@ -233,46 +233,68 @@ void NVPTXInstPrinter::printLdStCode(const MCInst *MI, int OpNum,
...
+        O << ".cta";
+        return;
+      case NVPTX::Scope::Cluster:
+        O << ".cluster";
+        return;
+      case NVPTX::Scope::Device:
+        O << ".gpu";
+        return;
       }
+      report_fatal_error(formatv(
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp:276]: The switch case is and should always be exhaustive, so doing that produces the following warning: 

```
/src/llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp:275:7: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
  275 |       default:
      |       ^
1 warning generated.
```

If a new scope is added and the now non-exhaustive switch does not trigger an error, the `fatal_error` after the switch unconditionally fails (otherwise it is dead code so should be removed).

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp [hunk 10]
```diff
@@ -233,46 +233,68 @@ void NVPTXInstPrinter::printLdStCode(const MCInst *MI, int OpNum,
...
       default:
         report_fatal_error(formatv(
-            "NVPTX LdStCode Printer does not support \"{}\" sem modifier.",
-            OrderingToCString(Ordering)));
+            "NVPTX LdStCode Printer does not support \"{}\" sem modifier. "
+            "Loads/Stores cannot be AcquireRelease or SequentiallyConsistent.",
+            OrderingToString(Ordering)));
+      }
+    } else if (!strcmp(Modifier, "sco")) {
+      auto S = NVPTX::Scope(Imm);
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp:259]: The variable is also used later on in the error message.

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp [hunk 11]
```diff
@@ -233,46 +233,68 @@ void NVPTXInstPrinter::printLdStCode(const MCInst *MI, int OpNum,
...
+        return;
       default:
         report_fatal_error(formatv(
-            "NVPTX LdStCode Printer does not support \"{}\" sem modifier.",
-            OrderingToCString(Ordering)));
+            "NVPTX LdStCode Printer does not support \"{}\" sem modifier. "
+            "Loads/Stores cannot be AcquireRelease or SequentiallyConsistent.",
+            OrderingToString(Ordering)));
+      }
+    } else if (!strcmp(Modifier, "sco")) {
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp]: Done

[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp]: Done

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp [hunk 12]
```diff
@@ -233,46 +233,68 @@ void NVPTXInstPrinter::printLdStCode(const MCInst *MI, int OpNum,
...
+        return;
       case NVPTX::PTXLdStInstCode::CONSTANT:
         O << ".const";
-        break;
+        return;
       case NVPTX::PTXLdStInstCode::GENERIC:
-        break;
+        return;
       default:
         llvm_unreachable("Wrong Address Space");
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp]: Done. Have added some printing utilities as well. There are more refactors for address spaces possible, like merging the two `enum`s of the backend, but i'd prefer to move those to a separate PR.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 3 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h (hunk 4 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:122]: Done

File: llvm/lib/Target/NVPTX/NVPTXSubtarget.h (hunk 7 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXSubtarget.h]: Done

File: llvm/lib/Target/NVPTX/NVPTX.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTX.h]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 13]
```diff
@@ -915,6 +920,42 @@ getOperationOrderings(MemSDNode *N, const NVPTXSubtarget *Subtarget) {
...
+    // with a smaller thread scope, we bump it here to system scope.
+    return NVPTX::Scope::System;
+  case NVPTX::Ordering::Relaxed:
+  case NVPTX::Ordering::Acquire:
+  case NVPTX::Ordering::Release:
+  case NVPTX::Ordering::AcquireRelease:
+  case NVPTX::Ordering::SequentiallyConsistent:
+    auto S = Scopes[N->getSyncScopeID()];
+
+    // Atomic operations must have a scope greater than thread.
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:948]: I made the scope change, and now "scope greater than thread" should make sense, but since thread is the smallest scope, i think `s == thread` is the right check. Even though the scope re-ordering doesn't help much here, it may still help in other places (if not now, in the future).

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 14]
```diff
@@ -957,33 +998,86 @@ static bool canLowerToLDG(MemSDNode *N, const NVPTXSubtarget &Subtarget,
...
+  case NVPTX::Ordering::AcquireRelease: {
+    switch (S) {
+    case NVPTX::Scope::System:
+      return T->hasMemoryOrdering() ? NVPTX::atomic_thread_fence_acq_rel_sys
+                                    : NVPTX::INT_MEMBAR_SYS;
+    case NVPTX::Scope::Block:
+      return T->hasMemoryOrdering() ? NVPTX::atomic_thread_fence_acq_rel_cta
+                                    : NVPTX::INT_MEMBAR_CTA;
+    case NVPTX::Scope::Cluster:
+      return NVPTX::atomic_thread_fence_acq_rel_cluster;
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1024]: There is already a check at the top of the function that checks that if the scope is cluster, they are supported.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 15]
```diff
@@ -957,33 +998,86 @@ static bool canLowerToLDG(MemSDNode *N, const NVPTXSubtarget &Subtarget,
...
+  case NVPTX::Ordering::SequentiallyConsistent: {
+    switch (S) {
+    case NVPTX::Scope::System:
+      return T->hasMemoryOrdering() ? NVPTX::atomic_thread_fence_seq_cst_sys
+                                    : NVPTX::INT_MEMBAR_SYS;
+    case NVPTX::Scope::Block:
+      return T->hasMemoryOrdering() ? NVPTX::atomic_thread_fence_seq_cst_cta
+                                    : NVPTX::INT_MEMBAR_CTA;
+    case NVPTX::Scope::Cluster:
+      return NVPTX::atomic_thread_fence_seq_cst_cluster;
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1043]: There is already a check at the top of the function that checks that if the scope is cluster, they are supported.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 16]
```diff
@@ -957,33 +998,86 @@ static bool canLowerToLDG(MemSDNode *N, const NVPTXSubtarget &Subtarget,
...
   }
   default:
     report_fatal_error(
         formatv("Unexpected fence ordering: \"{}\".",
-                OrderingToCString(NVPTX::Ordering(FenceOrdering))));
+                OrderingToString(NVPTX::Ordering(FenceOrdering))));
   }
-
-  return InstructionOrdering;
+  return std::make_pair(InstructionOrdering, Scope);
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1086]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 17]
```diff
@@ -915,6 +920,42 @@ getOperationOrderings(MemSDNode *N, const NVPTXSubtarget *Subtarget) {
...
+      report_fatal_error(
+          formatv("Atomics need scope > \"{}\".", ScopeToString(S)));
+
+    // If scope is cluster, clusters must be supported.
+    if (S == NVPTX::Scope::Cluster)
+      Subtarget->requireClusters("cluster scope");
+
+    // If operation is volatile, then its scope is system.
+    if (N->isVolatile())
+      S = NVPTX::Scope::System;
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Agreed, Done

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 1 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4198]: The IDs in the LLVM IR module are dynamic, and change depending on which order `syncscopes` appear in the module, so the numbers aren't really very useful in the error messages...

Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4198]: We already have a way to get the name: call this API `operator[]` which gives you a `NVPTX::Scope`, and then do `<<` on that, or `ScopeToString`. 

Within the error path of `operator[]`, we don't have a valid `NVPTX::Scope` yet, so have to do a bit more work here to get the name, and since we are in the error path, it may be that the ID isn't even in the map, so need to handle that case as well.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 2 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 6 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1289]: We can't use c++26 (I believe LLVM builds w/ c++17 at the moment).

The idea here is to make sure that the default vector size is sufficient to effectively accommodate expected uses.
Presumably previously we expected to see up to 12 elements. Now we've added another. While the array will fall back to the allocation, we do not want that. If we know we may need up to N elements, we should just say so.

There's no need to add any runtime checks, all you need to do is check the code below and see what's the largest number of additional elements we add to the array, and adjust the array size accordingly.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 1 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4198]: Perhaps it would make sense to extend Context API and query it for the scope name directly, instead of implementing the search here.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h (hunk 4 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:122]: The comment gives an impression that the fence instructions are always inserted, while IIUIC, it inserts fence instructions only *if* such instructions are needed.
I'd rephrase it along the lines of "Inserts an appropriate fence instruction, if it's required to implement particular ordering"

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 1 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4198]: Not sure what you mean. The scope discovery and map creation is done elsewhere. 

This is the function that maps a SelectionDag scope id to a NVPTX backend scope id.

AFAICT, any code anywhere can go through the Selection DAG, take a valid a scope id, and replace it with an invalid one, any time (e.g. after we build the map), for no good reason (e.g. a bug somewhere). 

This code is just doing a bounds check on that and trying to provide slightly better error reporting than just "you tried to index a map out of bounds somewhere".

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h (hunk 4 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.h:122]: Have improved the comment a bit.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 6 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1289]: I checked and we are not exceeding 12 elements here in any branch.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 18]
```diff
@@ -915,6 +926,39 @@ getOperationOrderings(MemSDNode *N, const NVPTXSubtarget *Subtarget) {
...
+      report_fatal_error(
+          formatv("Atomics need scope > \"{}\".", ScopeToString(S)));
+
+    // If scope is cluster, clusters must be supported.
+    if (S == NVPTX::Scope::Cluster)
+      Subtarget->failIfClustersUnsupported("cluster scope");
+
+    // If operation is volatile, then its scope is system.
+    return N->isVolatile() ? NVPTX::Scope::System : S;
+  }
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:959]: I'd add `llvm_unreachable()` here.

**Review by Artem-B — COMMENTED** [review]
Patch looks OK overall, modulo a few cosmetic improvements that could be extracted into separate patches.

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp [hunk 19]
```diff
@@ -95,228 +96,262 @@ void NVPTXInstPrinter::printOperand(const MCInst *MI, unsigned OpNo,
 }
 
 void NVPTXInstPrinter::printCvtMode(const MCInst *MI, int OpNum, raw_ostream &O,
-                                    const char *Modifier) {
+                                    const char *M) {
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp:99]: I think I misunderstood the request to use `StringRef`, to mean that those changes had to be part of this PR (my preference would have been to do those in different PRs..).

I didn't change the function signature in this PR because that would have pulled in a bit more changes into this PR (and its not straightforward due to how these APIs interact). My preference is to do that in a different PR.

File: llvm/test/CodeGen/NVPTX/load-store-sm-90.ll [hunk 20]
```diff
@@ -0,0 +1,1423 @@
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_90 -mattr=+ptx78 | FileCheck %s
+; RUN: %if ptxas-12.2 %{ llc < %s -march=nvptx64 -mcpu=sm_90 -mattr=+ptx78 | %ptxas-verify -arch=sm_90 %}
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply llvm/test/CodeGen/NVPTX/load-store-sm-90.ll:2]: Yes, I've run all of the tests with ptxas enabled (using CUDA 12.6.0) and they all passed.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 1 above)
Comment (gonzalobg) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:4198]: I see what you mean now. 
Yes, if the LLVM IR `SyncScope::ID` (in the LLVM syncscope order) is out of bounds of our cached map, we can't get the `NVPTX::Scope` id, so we error.

This may be, because there is a bug somewhere and either the `LLVM SyncScope` was incorrectly updated after we cached it, or an `ID` in the dag was incorrectly modified such that it triggered an out of bounds. 

In this error path, we are currently asking LLVM to give us all the SyncScope names. We are not copying the names themselves (i.e. the Strings), but references to the names into a `SmallVector<StringRef>`, and then we are checking if the `ID` actually is in bounds of that. If it is, we can print a name for it (and this may mean that the mapping was updated after we cached it), and if it is not in bounds, the ID was most likely messed up with.

A `getScopeName(ID)` API to the context would avoid the creation of the vector of string references.

Since this only impacts the path before we crash, would it be ok for me to add that API to context in a separate PR and as part of that PR update this to use it?  (More people may have an opinion on the Context API than just the backend, and I don't want to block this change on that).

**Comment by gonzalobg** [comment]
Have added the two `llvm_unreachable`.

For different sub-sequent PRs:
- There are many TODOs in the tests, that will get incrementally fixed over time (with smaller scoped changes).
- I should improve the Printer APIs to use `StringRef`.
- I should add a `getSyncScope(ID)` API to context, to get a single `StringRef`.
- I should explore merging the two `AddressSpace` enums into one.

File: llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/MCTargetDesc/NVPTXInstPrinter.cpp:99]: > my preference would have been to do those in different PRs..

+1 to splitting off independent changes into separate patches. Send them my way.

File: llvm/test/CodeGen/NVPTX/load-store-sm-90.ll (hunk 20 above)
[the remark this answers is not in the document]
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/load-store-sm-90.ll:2]: Awesome. Thank you!.

**Review by Artem-B — APPROVED** [review]

**Comment by gonzalobg** [comment]
Ping to merge (I still can't). Would be good to merge this early in the week in case there are any regressions.
